File: Services/AngleCalibrationCaptureService.vb
Imports System.Diagnostics
Imports System.Runtime.Versioning
Imports System.Threading
Imports WpfApp1.Models

Namespace Services
    <SupportedOSPlatform("windows")>
    Public Class AngleCalibrationCaptureService
        Implements IDisposable

        Private Const RenderFrameBufferCapacity As Integer = 128
        Private Const DefaultPacketBufferCapacity As Integer = 16384
        Private Const DefaultRenderFrameGenerationIntervalMilliseconds As Double = 1000.0 / 60.0

        Private ReadOnly _syncRoot As New Object()
        Private ReadOnly _queueProcessingSyncRoot As New Object()
        Private ReadOnly _rawInputBroker As IRawInputBroker
        Private ReadOnly _sessionRouter As RawMouseSessionRouter
        Private ReadOnly _engine As AngleCalibrationEngine
        Private ReadOnly _renderFrameBuffer As AngleCalibrationRenderFrameRingBuffer
        Private ReadOnly _renderFrameTimer As Timer
        Private ReadOnly _packetInbox As BufferedRawMousePacketInbox
        Private ReadOnly _packetPumpThread As Thread
        Private _renderFrameGenerationIntervalMilliseconds As Double = DefaultRenderFrameGenerationIntervalMilliseconds
        Private _capturePumpsActive As Integer
        Private _acceptedSessionGeneration As Integer
        Private _captureActive As Integer
        Private _packetEnqueuesInFlight As Integer
        Private _pendingQueuedPacket As QueuedRawMousePacket
        Private _packetConsumerDelayMilliseconds As Integer
        Private _disposed As Boolean

        Public Event DevicesChanged As EventHandler
        Public Event SelectedDeviceDisconnected As EventHandler

        Public Sub New(rawInputBroker As IRawInputBroker)
            Me.New(rawInputBroker, DefaultPacketBufferCapacity, 0)
        End Sub

        Friend Sub New(rawInputBroker As IRawInputBroker, packetBufferCapacity As Integer, Optional packetConsumerDelayMilliseconds As Integer = 0)
            If rawInputBroker Is Nothing Then
                Throw New ArgumentNullException(NameOf(rawInputBroker))
            End If

            _rawInputBroker = rawInputBroker
            _sessionRouter = New RawMouseSessionRouter(_rawInputBroker, RawMouseSessionMode.FirstActiveDevice)
            _engine = New AngleCalibrationEngine()
            _renderFrameBuffer = New AngleCalibrationRenderFrameRingBuffer(RenderFrameBufferCapacity)
            _renderFrameTimer = New Timer(AddressOf OnRenderFrameTimerTick, Nothing, Timeout.Infinite, Timeout.Infinite)
            _packetInbox = New BufferedRawMousePacketInbox(Math.Max(1, packetBufferCapacity))
            _packetConsumerDelayMilliseconds = Math.Max(0, packetConsumerDelayMilliseconds)
            _packetPumpThread = New Thread(AddressOf PacketPumpThreadMain) With {
                .IsBackground = True,
                .Name = "AngleCalibrationPacketPump"
            }

            AddHandler _rawInputBroker.MouseDevicesChanged, AddressOf OnDevicesChanged
            AddHandler _sessionRouter.PacketCaptured, AddressOf OnPacketCaptured
            AddHandler _sessionRouter.SelectedDeviceDisconnected, AddressOf OnSelectedDeviceDisconnected

            _packetPumpThread.Start()
        End Sub

        Public Function GetDevices() As IReadOnlyList(Of RawMouseDeviceInfo)
            Return FilterPhysicalDevices(_rawInputBroker.GetMouseDevices())
        End Function

        Public Function HasAvailableMouseDevice() As Boolean
            Dim devices = GetDevices()
            Return devices IsNot Nothing AndAlso devices.Count > 0
        End Function

        Public ReadOnly Property CurrentSessionDeviceId As String
            Get
                Return _sessionRouter.CurrentDeviceId
            End Get
        End Property

        Public Function BeginSession() As Boolean
            If Not HasAvailableMouseDevice() Then
                Volatile.Write(_captureActive, 0)
                _packetInbox.Drain()
                Volatile.Write(_capturePumpsActive, 0)

                SyncLock _syncRoot
                    _engine.SetLocked(False, GetNowMilliseconds())
                End SyncLock

                UpdateRenderFrameTimer()
                CaptureRenderFrame()
                Return False
            End If

            Volatile.Write(_captureActive, 0)
            _packetInbox.Drain()

            _sessionRouter.BeginSession()
            Volatile.Write(_acceptedSessionGeneration, _sessionRouter.SessionGeneration)

            Dim nowMs = GetNowMilliseconds()
            SyncLock _syncRoot
                _engine.SetLocked(True, nowMs)
            End SyncLock

            Volatile.Write(_capturePumpsActive, 1)
            UpdateRenderFrameTimer()
            Volatile.Write(_captureActive, 1)
            CaptureRenderFrame()

            Return True
        End Function

        Public Sub PauseSession()
            _sessionRouter.PauseSession()

            WaitForPacketEnqueueToSettle()
            FlushQueuedPackets()

            Volatile.Write(_captureActive, 0)
            Volatile.Write(_acceptedSessionGeneration, _sessionRouter.SessionGeneration)
            _packetInbox.Drain()
            Volatile.Write(_capturePumpsActive, 0)

            Dim nowMs = GetNowMilliseconds()
            SyncLock _syncRoot
                _engine.SetLocked(False, nowMs)
            End SyncLock

            UpdateRenderFrameTimer()
            CaptureRenderFrame()
        End Sub

        Public Sub StopSession()
            _sessionRouter.StopSession()

            WaitForPacketEnqueueToSettle()
            FlushQueuedPackets()

            Volatile.Write(_captureActive, 0)
            Volatile.Write(_acceptedSessionGeneration, _sessionRouter.SessionGeneration)
            _packetInbox.Drain()
            Volatile.Write(_capturePumpsActive, 0)

            Dim nowMs = GetNowMilliseconds()
            SyncLock _syncRoot
                _engine.SetLocked(False, nowMs)
            End SyncLock

            UpdateRenderFrameTimer()
            CaptureRenderFrame()
        End Sub

        Public Sub ResetSession()
            Volatile.Write(_captureActive, 0)
            _packetInbox.Drain()
            Volatile.Write(_capturePumpsActive, 0)

            SyncLock _syncRoot
                _engine.Reset()
                _renderFrameBuffer.Clear()
            End SyncLock

            UpdateRenderFrameTimer()
            CaptureRenderFrame()
        End Sub

        Public Function CaptureRenderFrame() As AngleCalibrationRenderFrame
            SyncLock _syncRoot
                Return CaptureRenderFrameCore(GetNowMilliseconds())
            End SyncLock
        End Function

        Public Function TryReadLatestRenderFrame(ByRef renderFrame As AngleCalibrationRenderFrame) As Boolean
            SyncLock _syncRoot
                Return _renderFrameBuffer.TryReadLatest(renderFrame)
            End SyncLock
        End Function

        Private Function CaptureRenderFrameCore(nowMs As Double) As AngleCalibrationRenderFrame
            Dim renderFrame = _engine.CreateRenderFrame(nowMs)
            _renderFrameBuffer.Write(renderFrame)
            Return renderFrame
        End Function

        Private Sub OnDevicesChanged(sender As Object, e As EventArgs)
            RaiseEvent DevicesChanged(Me, EventArgs.Empty)
        End Sub

        Private Sub OnPacketCaptured(sender As Object, e As RawMousePacketEventArgs)
            If e Is Nothing OrElse e.Packet Is Nothing Then
                Return
            End If

            If Volatile.Read(_captureActive) = 0 Then
                Return
            End If

            Dim acceptedGeneration = Volatile.Read(_acceptedSessionGeneration)
            If e.SessionGeneration <> acceptedGeneration Then
                Return
            End If

            Interlocked.Increment(_packetEnqueuesInFlight)
            Try
                Dim droppedCount = 0
                _packetInbox.Enqueue(e.Packet, acceptedGeneration, droppedCount)
            Finally
                Interlocked.Decrement(_packetEnqueuesInFlight)
            End Try
        End Sub

        Private Sub OnSelectedDeviceDisconnected(sender As Object, e As EventArgs)
            WaitForPacketEnqueueToSettle()
            FlushQueuedPackets()

            Volatile.Write(_captureActive, 0)
            Volatile.Write(_acceptedSessionGeneration, _sessionRouter.SessionGeneration)
            _packetInbox.Drain()
            Volatile.Write(_capturePumpsActive, 0)

            Dim nowMs = GetNowMilliseconds()
            SyncLock _syncRoot
                _engine.SetLocked(False, nowMs)
            End SyncLock

            UpdateRenderFrameTimer()
            CaptureRenderFrame()
            RaiseEvent SelectedDeviceDisconnected(Me, EventArgs.Empty)
        End Sub

        Private Sub OnRenderFrameTimerTick(state As Object)
            If _disposed OrElse Volatile.Read(_capturePumpsActive) = 0 Then
                Return
            End If

            SyncLock _syncRoot
                CaptureRenderFrameCore(GetNowMilliseconds())
            End SyncLock
        End Sub

        Private Sub UpdateRenderFrameTimer()
            Dim dueTime = Timeout.InfiniteTimeSpan
            Dim period = Timeout.InfiniteTimeSpan

            If Volatile.Read(_capturePumpsActive) <> 0 Then
                period = TimeSpan.FromMilliseconds(_renderFrameGenerationIntervalMilliseconds)
                dueTime = TimeSpan.Zero
            End If

            _renderFrameTimer.Change(dueTime, period)
        End Sub

        Private Sub PacketPumpThreadMain()
            Do
                Dim queuedPacket As QueuedRawMousePacket = Nothing
                If Not _packetInbox.WaitDequeue(queuedPacket) Then
                    Exit Do
                End If

                Interlocked.Exchange(_pendingQueuedPacket, queuedPacket)

                SyncLock _queueProcessingSyncRoot
                    DrainQueuedPacketsUnsafe()
                End SyncLock
            Loop
        End Sub

        Private Sub FlushQueuedPackets()
            SyncLock _queueProcessingSyncRoot
                DrainQueuedPacketsUnsafe()
            End SyncLock
        End Sub

        Private Sub DrainQueuedPacketsUnsafe()
            Dim pendingQueuedPacket = Interlocked.Exchange(_pendingQueuedPacket, Nothing)
            If pendingQueuedPacket IsNot Nothing Then
                ProcessPacket(pendingQueuedPacket)
            End If

            Dim queuedPacket As QueuedRawMousePacket = Nothing
            While _packetInbox.TryDequeue(queuedPacket)
                ProcessPacket(queuedPacket)
            End While
        End Sub

        Private Sub ProcessPacket(queuedPacket As QueuedRawMousePacket)
            If queuedPacket Is Nothing OrElse queuedPacket.Packet Is Nothing Then
                Return
            End If

            If _packetConsumerDelayMilliseconds > 0 Then
                Thread.Sleep(_packetConsumerDelayMilliseconds)
            End If

            SyncLock _syncRoot
                If Volatile.Read(_captureActive) = 0 Then
                    Return
                End If

                If queuedPacket.Generation <> Volatile.Read(_acceptedSessionGeneration) Then
                    Return
                End If

                _engine.PushPacket(queuedPacket.Packet)
            End SyncLock
        End Sub

        Private Shared Function GetNowMilliseconds() As Double
            Return Stopwatch.GetTimestamp() * 1000.0 / Stopwatch.Frequency
        End Function

        Private Sub WaitForPacketEnqueueToSettle(Optional timeoutMilliseconds As Integer = 1000)
            Dim timeout = Math.Max(0, timeoutMilliseconds)
            Dim deadline = Environment.TickCount64 + timeout

            While Volatile.Read(_packetEnqueuesInFlight) > 0
                If timeout > 0 AndAlso Environment.TickCount64 >= deadline Then
                    Exit While
                End If

                Thread.Sleep(1)
            End While
        End Sub

        Public Sub Dispose() Implements IDisposable.Dispose
            If _disposed Then
                Return
            End If

            _disposed = True
            Volatile.Write(_captureActive, 0)
            Volatile.Write(_capturePumpsActive, 0)

            _renderFrameTimer.Change(Timeout.Infinite, Timeout.Infinite)
            _renderFrameTimer.Dispose()

            _packetInbox.Dispose()
            If _packetPumpThread IsNot Nothing AndAlso _packetPumpThread.IsAlive Then
                _packetPumpThread.Join(1000)
            End If

            RemoveHandler _rawInputBroker.MouseDevicesChanged, AddressOf OnDevicesChanged
            RemoveHandler _sessionRouter.PacketCaptured, AddressOf OnPacketCaptured
            RemoveHandler _sessionRouter.SelectedDeviceDisconnected, AddressOf OnSelectedDeviceDisconnected
            _sessionRouter.Dispose()
        End Sub
    End Class
End Namespace

File: src/ClickSyncMouseTester/Services/AngleCalibrationCaptureService.cs
using ClickSyncMouseTester.Models;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Runtime.Versioning;
using System.Threading;

namespace ClickSyncMouseTester.Services;

[SupportedOSPlatform("windows")]
public class AngleCalibrationCaptureService : IDisposable
{
    private const int RenderFrameBufferCapacity = 128;

    private const int DefaultPacketBufferCapacity = 65536;

    private const int PacketPumpShutdownJoinTimeoutMilliseconds = 1000;

    private const double DefaultRenderFrameGenerationIntervalMilliseconds = 16.666666666666668;

    private readonly object _syncRoot;

    private readonly object _queueProcessingSyncRoot;

    private readonly IRawMouseReportSource _mouseReportSource;

    private readonly IRawInputDeviceCatalog _deviceCatalog;

    private readonly RawMouseSessionRouter _sessionRouter;

    private readonly AngleCalibrationEngine _engine;

    private readonly AngleCalibrationRenderFrameRingBuffer _renderFrameBuffer;

    private readonly Timer _renderFrameTimer;

    private readonly BufferedRawMousePacketInbox _packetInbox;

    private readonly Thread _packetPumpThread;

    private double _renderFrameGenerationIntervalMilliseconds;

    private int _capturePumpsActive;

    private int _acceptedSessionGeneration;

    private int _captureActive;

    private int _packetEnqueuesInFlight;

    private int _packetConsumerDelayMilliseconds;

    private int _droppedPacketCount;

    private bool _disposed;

    public string CurrentSessionDeviceId => _sessionRouter.CurrentDeviceId;

    public int DroppedPacketCount => Math.Max(0, Interlocked.Add(ref _droppedPacketCount, 0));

    public event EventHandler DevicesChanged;

    public event EventHandler SelectedDeviceDisconnected;

    public AngleCalibrationCaptureService(IRawInputBroker rawInputBroker)
        : this(rawInputBroker, DefaultPacketBufferCapacity)
    {
    }

    internal AngleCalibrationCaptureService(IRawInputBroker rawInputBroker, int packetBufferCapacity, int packetConsumerDelayMilliseconds = 0)
        : this(rawInputBroker, rawInputBroker, packetBufferCapacity, packetConsumerDelayMilliseconds)
    {
    }

    internal AngleCalibrationCaptureService(IRawMouseReportSource mouseReportSource, IRawInputDeviceCatalog deviceCatalog, int packetBufferCapacity, int packetConsumerDelayMilliseconds = 0)
    {
        _syncRoot = new object();
        _queueProcessingSyncRoot = new object();
        _renderFrameGenerationIntervalMilliseconds = 16.666666666666668;
        if (mouseReportSource == null)
        {
            throw new ArgumentNullException(nameof(mouseReportSource));
        }
        if (deviceCatalog == null)
        {
            throw new ArgumentNullException(nameof(deviceCatalog));
        }
        _mouseReportSource = mouseReportSource;
        _deviceCatalog = deviceCatalog;
        _sessionRouter = new RawMouseSessionRouter(_mouseReportSource, _deviceCatalog, RawMouseSessionMode.FirstActiveDevice);
        _engine = new AngleCalibrationEngine();
        _renderFrameBuffer = new AngleCalibrationRenderFrameRingBuffer(128);
        _renderFrameTimer = new Timer(OnRenderFrameTimerTick, null, -1, -1);
        _packetInbox = new BufferedRawMousePacketInbox(Math.Max(1, packetBufferCapacity));
        _packetConsumerDelayMilliseconds = Math.Max(0, packetConsumerDelayMilliseconds);
        _packetPumpThread = new Thread(PacketPumpThreadMain)
        {
            IsBackground = true,
            Name = "AngleCalibrationPacketPump"
        };
        _deviceCatalog.MouseDevicesChanged += OnDevicesChanged;
        _sessionRouter.PacketCaptured += OnPacketCaptured;
        _sessionRouter.SelectedDeviceDisconnected += OnSelectedDeviceDisconnected;
        _packetPumpThread.Start();
    }

    public IReadOnlyList<RawMouseDeviceInfo> GetDevices()
    {
        return MouseDeviceFiltering.FilterSelectableMotionDevices(_deviceCatalog.GetMouseDevices(), _deviceCatalog.GetMouseEndpointActivitySnapshots());
    }

    public bool HasAvailableMouseDevice()
    {
        IReadOnlyList<RawMouseDeviceInfo> devices = GetDevices();
        if (devices != null)
        {
            return devices.Count > 0;
        }
        return false;
    }

    public bool BeginSession()
    {
        if (!HasAvailableMouseDevice())
        {
            Volatile.Write(ref _captureActive, 0);
            _packetInbox.Drain();
            Volatile.Write(ref _capturePumpsActive, 0);
            object syncRoot = _syncRoot;
            bool lockTaken = false;
            try
            {
                Monitor.Enter(syncRoot, ref lockTaken);
                _engine.SetLocked(isLocked: false, GetNowMilliseconds());
            }
            finally
            {
                if (lockTaken)
                {
                    Monitor.Exit(syncRoot);
                }
            }
            UpdateRenderFrameTimer();
            CaptureRenderFrame();
            return false;
        }
        Volatile.Write(ref _captureActive, 0);
        _packetInbox.Drain();
        _sessionRouter.BeginSession();
        Volatile.Write(ref _acceptedSessionGeneration, _sessionRouter.SessionGeneration);
        double nowMilliseconds = GetNowMilliseconds();
        object syncRoot2 = _syncRoot;
        bool lockTaken2 = false;
        try
        {
            Monitor.Enter(syncRoot2, ref lockTaken2);
            _engine.SetLocked(isLocked: true, nowMilliseconds);
        }
        finally
        {
            if (lockTaken2)
            {
                Monitor.Exit(syncRoot2);
            }
        }
        Volatile.Write(ref _capturePumpsActive, 1);
        UpdateRenderFrameTimer();
        Volatile.Write(ref _captureActive, 1);
        CaptureRenderFrame();
        return true;
    }

    public void PauseSession()
    {
        _sessionRouter.PauseSession();
        WaitForPacketEnqueueToSettle();
        FlushQueuedPackets();
        Volatile.Write(ref _captureActive, 0);
        Volatile.Write(ref _acceptedSessionGeneration, _sessionRouter.SessionGeneration);
        _packetInbox.Drain();
        Volatile.Write(ref _capturePumpsActive, 0);
        double nowMilliseconds = GetNowMilliseconds();
        object syncRoot = _syncRoot;
        bool lockTaken = false;
        try
        {
            Monitor.Enter(syncRoot, ref lockTaken);
            _engine.SetLocked(isLocked: false, nowMilliseconds);
        }
        finally
        {
            if (lockTaken)
            {
                Monitor.Exit(syncRoot);
            }
        }
        UpdateRenderFrameTimer();
        CaptureRenderFrame();
    }

    public void StopSession()
    {
        _sessionRouter.StopSession();
        WaitForPacketEnqueueToSettle();
        FlushQueuedPackets();
        Volatile.Write(ref _captureActive, 0);
        Volatile.Write(ref _acceptedSessionGeneration, _sessionRouter.SessionGeneration);
        _packetInbox.Drain();
        Volatile.Write(ref _capturePumpsActive, 0);
        double nowMilliseconds = GetNowMilliseconds();
        object syncRoot = _syncRoot;
        bool lockTaken = false;
        try
        {
            Monitor.Enter(syncRoot, ref lockTaken);
            _engine.SetLocked(isLocked: false, nowMilliseconds);
        }
        finally
        {
            if (lockTaken)
            {
                Monitor.Exit(syncRoot);
            }
        }
        UpdateRenderFrameTimer();
        CaptureRenderFrame();
    }

    public void ResetSession()
    {
        Volatile.Write(ref _captureActive, 0);
        WaitForPacketEnqueueToSettle();
        object queueProcessingSyncRoot = _queueProcessingSyncRoot;
        bool lockTaken = false;
        try
        {
            Monitor.Enter(queueProcessingSyncRoot, ref lockTaken);
            _packetInbox.Drain();
            Volatile.Write(ref _capturePumpsActive, 0);
            Interlocked.Exchange(ref _droppedPacketCount, 0);
            object syncRoot = _syncRoot;
            bool lockTaken2 = false;
            try
            {
                Monitor.Enter(syncRoot, ref lockTaken2);
                _engine.Reset();
                _renderFrameBuffer.Clear();
            }
            finally
            {
                if (lockTaken2)
                {
                    Monitor.Exit(syncRoot);
                }
            }
        }
        finally
        {
            if (lockTaken)
            {
                Monitor.Exit(queueProcessingSyncRoot);
            }
        }
        UpdateRenderFrameTimer();
        CaptureRenderFrame();
    }

    public AngleCalibrationRenderFrame CaptureRenderFrame()
    {
        object syncRoot = _syncRoot;
        bool lockTaken = false;
        try
        {
            Monitor.Enter(syncRoot, ref lockTaken);
            return CaptureRenderFrameCore(GetNowMilliseconds());
        }
        finally
        {
            if (lockTaken)
            {
                Monitor.Exit(syncRoot);
            }
        }
    }

    public bool TryReadLatestRenderFrame(ref AngleCalibrationRenderFrame renderFrame)
    {
        object syncRoot = _syncRoot;
        bool lockTaken = false;
        try
        {
            Monitor.Enter(syncRoot, ref lockTaken);
            return _renderFrameBuffer.TryReadLatest(ref renderFrame);
        }
        finally
        {
            if (lockTaken)
            {
                Monitor.Exit(syncRoot);
            }
        }
    }

    private AngleCalibrationRenderFrame CaptureRenderFrameCore(double nowMs)
    {
        AngleCalibrationRenderFrame angleCalibrationRenderFrame = _engine.CreateRenderFrame(nowMs);
        _renderFrameBuffer.Write(angleCalibrationRenderFrame);
        return angleCalibrationRenderFrame;
    }

    private void OnDevicesChanged(object sender, EventArgs e)
    {
        DevicesChanged?.Invoke(this, EventArgs.Empty);
    }

    private void OnPacketCaptured(object sender, RawMousePacketEventArgs e)
    {
        if (e == null || e.Packet == null)
        {
            return;
        }
        Interlocked.Increment(ref _packetEnqueuesInFlight);
        try
        {
            if (Volatile.Read(in _captureActive) == 0)
            {
                return;
            }
            int acceptedSessionGeneration = Volatile.Read(ref _acceptedSessionGeneration);
            if (e.SessionGeneration == acceptedSessionGeneration)
            {
                int droppedCount = 0;
                _packetInbox.Enqueue(e.Packet, acceptedSessionGeneration, ref droppedCount);
                if (droppedCount > 0)
                {
                    Interlocked.Add(ref _droppedPacketCount, droppedCount);
                }
            }
        }
        finally
        {
            Interlocked.Decrement(ref _packetEnqueuesInFlight);
        }
    }

    private void OnSelectedDeviceDisconnected(object sender, EventArgs e)
    {
        WaitForPacketEnqueueToSettle();
        FlushQueuedPackets();
        Volatile.Write(ref _captureActive, 0);
        Volatile.Write(ref _acceptedSessionGeneration, _sessionRouter.SessionGeneration);
        _packetInbox.Drain();
        Volatile.Write(ref _capturePumpsActive, 0);
        double nowMilliseconds = GetNowMilliseconds();
        object syncRoot = _syncRoot;
        bool lockTaken = false;
        try
        {
            Monitor.Enter(syncRoot, ref lockTaken);
            _engine.SetLocked(isLocked: false, nowMilliseconds);
        }
        finally
        {
            if (lockTaken)
            {
                Monitor.Exit(syncRoot);
            }
        }
        UpdateRenderFrameTimer();
        CaptureRenderFrame();
        SelectedDeviceDisconnected?.Invoke(this, EventArgs.Empty);
    }

    private void OnRenderFrameTimerTick(object state)
    {
        if (_disposed || Volatile.Read(in _capturePumpsActive) == 0)
        {
            return;
        }
        object syncRoot = _syncRoot;
        bool lockTaken = false;
        try
        {
            Monitor.Enter(syncRoot, ref lockTaken);
            CaptureRenderFrameCore(GetNowMilliseconds());
        }
        finally
        {
            if (lockTaken)
            {
                Monitor.Exit(syncRoot);
            }
        }
    }

    private void UpdateRenderFrameTimer()
    {
        TimeSpan dueTime = Timeout.InfiniteTimeSpan;
        TimeSpan period = Timeout.InfiniteTimeSpan;
        if (Volatile.Read(in _capturePumpsActive) != 0)
        {
            period = TimeSpan.FromMilliseconds(_renderFrameGenerationIntervalMilliseconds);
            dueTime = TimeSpan.Zero;
        }
        _renderFrameTimer.Change(dueTime, period);
    }

    private void PacketPumpThreadMain()
    {
        while (_packetInbox.WaitForData())
        {
            object queueProcessingSyncRoot = _queueProcessingSyncRoot;
            bool lockTaken = false;
            try
            {
                Monitor.Enter(queueProcessingSyncRoot, ref lockTaken);
                DrainQueuedPacketsUnsafe();
            }
            finally
            {
                if (lockTaken)
                {
                    Monitor.Exit(queueProcessingSyncRoot);
                }
            }
        }
    }

    private void FlushQueuedPackets()
    {
        object queueProcessingSyncRoot = _queueProcessingSyncRoot;
        bool lockTaken = false;
        try
        {
            Monitor.Enter(queueProcessingSyncRoot, ref lockTaken);
            DrainQueuedPacketsUnsafe();
        }
        finally
        {
            if (lockTaken)
            {
                Monitor.Exit(queueProcessingSyncRoot);
            }
        }
    }

    private void DrainQueuedPacketsUnsafe()
    {
        QueuedRawMousePacket queuedPacket = QueuedRawMousePacket.Empty;
        while (_packetInbox.TryDequeue(ref queuedPacket))
        {
            ProcessPacket(queuedPacket);
        }
    }

    private void ProcessPacket(QueuedRawMousePacket queuedPacket)
    {
        if (!queuedPacket.HasPacket || Volatile.Read(in _captureActive) == 0 || queuedPacket.Generation != Volatile.Read(in _acceptedSessionGeneration))
        {
            return;
        }
        if (_packetConsumerDelayMilliseconds > 0)
        {
            Thread.Sleep(_packetConsumerDelayMilliseconds);
        }
        object syncRoot = _syncRoot;
        bool lockTaken = false;
        try
        {
            Monitor.Enter(syncRoot, ref lockTaken);
            if (Volatile.Read(in _captureActive) != 0 && queuedPacket.Generation == Volatile.Read(in _acceptedSessionGeneration))
            {
                _engine.PushPacket(queuedPacket.Packet);
            }
        }
        finally
        {
            if (lockTaken)
            {
                Monitor.Exit(syncRoot);
            }
        }
    }

    private static double GetNowMilliseconds()
    {
        return (double)Stopwatch.GetTimestamp() * 1000.0 / (double)Stopwatch.Frequency;
    }

    private void WaitForPacketEnqueueToSettle(int timeoutMilliseconds = 1000)
    {
        int timeout = Math.Max(0, timeoutMilliseconds);
        long deadlineTick = Environment.TickCount64 + timeout;
        while (Volatile.Read(ref _packetEnqueuesInFlight) > 0 && (timeout <= 0 || Environment.TickCount64 < deadlineTick))
        {
            Thread.Sleep(1);
        }
    }

    public void Dispose()
    {
        if (!_disposed)
        {
            _disposed = true;
            Volatile.Write(ref _captureActive, 0);
            Volatile.Write(ref _capturePumpsActive, 0);
            _packetConsumerDelayMilliseconds = 0;
            _renderFrameTimer.Change(-1, -1);
            _renderFrameTimer.Dispose();
            _packetInbox.Complete();
            bool packetPumpStopped = true;
            if (_packetPumpThread != null && _packetPumpThread.IsAlive)
            {
                packetPumpStopped = _packetPumpThread.Join(PacketPumpShutdownJoinTimeoutMilliseconds);
            }
            if (packetPumpStopped)
            {
                _packetInbox.Dispose();
            }
            _deviceCatalog.MouseDevicesChanged -= OnDevicesChanged;
            _sessionRouter.PacketCaptured -= OnPacketCaptured;
            _sessionRouter.SelectedDeviceDisconnected -= OnSelectedDeviceDisconnected;
            _sessionRouter.Dispose();
        }
    }

    void IDisposable.Dispose()
    {
        this.Dispose();
    }
}





